File: modules/_foundation/src/main/java/com/projecthub/base/project/domain/event/ProjectUpdatedEvent.java
package com.projecthub.base.project.domain.event;

import com.projecthub.base.project.domain.entity.Project;

import java.time.Instant;
import java.util.Map;
import java.util.UUID;

public record ProjectUpdatedEvent(
    UUID eventId,
    UUID projectId,
    String projectName,
    Instant timestamp,
    UUID initiatorId,
    Map<String, Object> updatedFields
) implements ProjectEvent {

    public ProjectUpdatedEvent(final Project project, final UUID initiatorId, final Map<String, Object> updatedFields) {
        this(
            UUID.randomUUID(),
            project.getId(),
            project.getName(),
            Instant.now(),
            initiatorId,
            updatedFields
        );
    }

    @Override
    public UUID getEventId() {
        return this.eventId;
    }

    @Override
    public UUID getProjectId() {
        return this.projectId;
    }

    @Override
    public Instant getOccurredOn() {
        return this.timestamp;
    }

    @Override
    public String getProjectName() {
        return projectName();
    }
}

File: src/main/kotlin/com/projecthub/domain/project/events/ProjectUpdatedEvent.kt
package com.projecthub.domain.project.events

import com.projecthub.events.DomainEvent
import java.time.Instant
import java.util.UUID

/**
 * Event representing project update.
 */
class ProjectUpdatedEvent(
    override val aggregateId: UUID,
    val name: String,
    val description: String,
    val startDate: Instant,
    val dueDate: Instant? = null
) : DomainEvent() {
    override val type: String = "ProjectUpdatedEvent"
}
